Space closing

odzioo123 opened the pull request.

odzioo123 committed:
Added backend part for closing space.
atlas_main - fn delete_space, and it calls to atlas_space - fn clean_up_space_before_deletion.

odzioo123 committed:
Added frontend for deleting space + small fix to backend

odzioo123 committed:
cargo fmt + small fix

wojciechp6 commented on src/atlas_space/atlas_space.did:
```diff
@@ -73,9 +74,10 @@ type GetTasksRes = record {
-type Result_1 = variant { Ok : nat64; Err : Error };
-type Result_2 = variant { Ok : GetClosedTasksRes; Err : Error };
-type Result_3 = variant { Ok : GetTasksRes; Err : Error };
+type Result_1 = variant { Ok; Err : text };
```
Why you droped the type?

odzioo123 replied:
It just changed the order automatically in did-s

wojciechp6 commented on src/atlas_space/atlas_space.did:
```diff
@@ -73,9 +74,10 @@ type GetTasksRes = record {
-type Result_2 = variant { Ok : GetClosedTasksRes; Err : Error };
-type Result_3 = variant { Ok : GetTasksRes; Err : Error };
+type Result_1 = variant { Ok; Err : text };
+type Result_2 = variant { Ok : nat64; Err : Error };
+type Result_3 = variant { Ok : GetClosedTasksRes; Err : Error };
+type Result_4 = variant { Ok : GetTasksRes; Err : Error };
```
I would propose to name this types in some better way

odzioo123 replied:
These are just rust Result types, it is generated for example from this method:
```rust
pub fn get_open_tasks(args: GetTasksArgs) -> Result<GetTasksRes, Error>
```

I don't think I should change them

wojciechp6 commented on src/atlas_space/src/methods/update.rs:
```diff
@@ -175,3 +176,36 @@ pub fn transfer_space(to: Principal) {
 
     memory::mut_config(|config| config.owner = to);
 }
+
+#[update]
+pub async fn clean_up_space_before_deletion() -> Result<(), String> {
```
Why isn't it just `close_space`?

odzioo123 replied:
Because it describes better what does it do. This one doesn't really delete space, just clean it for deletion.

wojciechp6 replied:
So, is there any function to delete the space with all the content?

wojciechp6 commented on src/declarations/atlas_main/atlas_main.did.d.ts:
```diff
@@ -63,9 +67,9 @@ export type Rank = { 'SpaceLead' : null } |
   { 'Admin' : null };
 export type Result = { 'Ok' : Space } |
   { 'Err' : Error };
-export type Result_1 = { 'Ok' : GetSpacesRes } |
+export type Result_1 = { 'Ok' : null } |
   { 'Err' : Error };
-export type Result_2 = { 'Ok' : null } |
+export type Result_2 = { 'Ok' : GetSpacesRes } |
```
Again, name this types

wojciechp6 commented on src/declarations/atlas_space/atlas_space.did:
```diff
@@ -73,9 +74,10 @@ type GetTasksRes = record {
-type Result_1 = variant { Ok : nat64; Err : Error };
-type Result_2 = variant { Ok : GetClosedTasksRes; Err : Error };
-type Result_3 = variant { Ok : GetTasksRes; Err : Error };
+type Result_1 = variant { Ok; Err : text };
+type Result_2 = variant { Ok : nat64; Err : Error };
+type Result_3 = variant { Ok : GetClosedTasksRes; Err : Error };
+type Result_4 = variant { Ok : GetTasksRes; Err : Error };
```
Name it

odzioo123 replied:
These are just rust Result types, it is generated for example from this method:
```rust
pub fn get_open_tasks(args: GetTasksArgs) -> Result<GetTasksRes, Error>
```

I don't think I should change them

wojciechp6 commented on src/declarations/atlas_space/atlas_space.did.js:
```diff
@@ -18,38 +19,42 @@ export const idlFactory = ({ IDL }) => {
+      'ConfigNotSet' : IDL.Null,
+      'UserAlreadyRewarded' : IDL.Null,
+      'TaskNotFound' : IDL.Nat64,
+      'FailedToClaimRewards' : IDL.Vec(IDL.Tuple(IDL.Nat64, Error)),
```
Maybe extract type?

odzioo123 replied:
I don't really understand. (did files are generated automatically from types and methods exported by canister)

wojciechp6 commented on src/atlas_space/src/methods/update.rs:
```diff
@@ -175,3 +176,36 @@ pub fn transfer_space(to: Principal) {
+            "Failed to claim rewards for {} tasks: {:?}",
+            errors.len(),
+            errors
+        ));
```
Try to make the update functions as clear as possible. Extract and name subfunctions.

wojciechp6 reviewed (commented):
Try to make functions short and clear.

Make sure to merge it properly - conflicts with other PR.

odzioo123 committed:
small fix to rebase

pu0238 commented on src/atlas_main/src/memory.rs:
```diff
@@ -128,6 +128,30 @@ pub fn push_space(space_principal: &Space) -> Result<(), Error> {
+            new_stable_vec
+                .push(space)
+                .map_err(|err| Error::FailedToSaveSpace(format!("{err:?}")))?;
+        }
```
Not ike this just get mut vector from SPACES_VEC and set it to temp_vec.iter().colect()

pu0238 replied:
My mistake — we actually should not remove the space. Removing a space would change its index, and we rely on that index to bind owners to the spaces they own. So we must avoid doing that.

Instead, change the vector type to use an Option. That way, we can preserve the indexes as they are. After making this change, you'll also need to create a migration to upgrade the existing data accordingly.

pu0238 commented on src/atlas_main/src/methods/update.rs:
```diff
@@ -263,3 +267,64 @@ pub async fn transfer_space(args: TransferSpace) -> Result<(), Error> {
+            } else {
+                None
+            }
+        })
```
use `is_some_and` 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=aa196c2b7e60afb18c9199b9f9067bc4
https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some_and

pu0238 commented on src/atlas_main/src/methods/update.rs:
```diff
@@ -263,3 +267,64 @@ pub async fn transfer_space(args: TransferSpace) -> Result<(), Error> {
+        })
+        .ok_or(Error::UserNotOwner)?;
+
+    let cleanup_result: Result<(), String> =
+        Call::bounded_wait(space_id, "clean_up_space_before_deletion")
+            .with_arg(())
+            .await
+            .map_err(|err| Error::FailedToCallSpace {
+                err: err.to_string(),
+                principal: space_id,
+            })?
+            .candid::<Result<(), String>>()
+            .map_err(|err| Error::FailedToCleanSpace {
+                err: err.to_string(),
+                principal: space_id,
+            })?;
```
we should panic here no to just rerur error. if something was wrong here we should not let it be saved

pu0238 commented on src/atlas_main/src/methods/update.rs:
```diff
@@ -263,3 +267,64 @@ pub async fn transfer_space(args: TransferSpace) -> Result<(), Error> {
+        canister_id: space_id,
+    })
+    .await
+    .map_err(|err| Error::FailedToDeleteCanister(err.to_string()))?;
```
we also should panic here and provent any thing that happend before

pu0238 commented on src/atlas_main/src/methods/update.rs:
```diff
@@ -263,3 +267,64 @@ pub async fn transfer_space(args: TransferSpace) -> Result<(), Error> {
+        principal: space_id,
+    })?;
+
+    stop_canister(&StopCanisterArgs {
```
we also should panic here and provent any thing that happend before

odzioo123 replied:
if code got to delete_canister or to stop_canister and panicked/returned error, we probably need to add something to just retry this part, but I am not sure.

pu0238 commented on src/atlas_space/src/methods/update.rs:
```diff
@@ -175,3 +176,36 @@ pub fn transfer_space(to: Principal) {
+pub async fn clean_up_space_before_deletion() -> Result<(), String> {
+    parent_guard().map_err(|e| e.to_string())?;
+
+    let open_tasks: Vec<_> = memory::with_open_tasks_iter(|iter| iter.collect());
```
just create getter for that iter is less efficent in here. create getter function in memory

pu0238 commented on src/atlas_space/src/methods/update.rs:
```diff
@@ -175,3 +176,36 @@ pub fn transfer_space(to: Principal) {
+            .map_err(|e| format!("Failed to close task {task_id:?}: {e:?}"))?;
+    }
+
+    let mut errors = vec![];
```
let mut errors = Vec::new();

pu0238 commented on src/atlas_space/src/methods/update.rs:
```diff
@@ -175,3 +176,36 @@ pub fn transfer_space(to: Principal) {
+
+    let mut errors = vec![];
+    let closed_tasks: Vec<(TaskId, ClosedTask)> =
+        memory::with_closed_tasks_iter(|iter| iter.collect());
```
ditto

pu0238 commented on src/atlas_space/src/tasks/closed_task.rs:
```diff
@@ -55,6 +56,21 @@ impl ClosedTask {
+            .tasks
+            .iter()
+            .flat_map(|task| task.get_submission_map().keys().cloned())
+            .collect();
```
just use filter in here to find accepted submitions and also check in the same loop did they are claimed. ther witrdraw them

odzioo123 committed:
claim_all_rewards works as claim_reward.

changed SPACES_VEC to store Option<Space> instead of <Space>.

added migration for SPACES_VEC.

minor logic fixes

odzioo123 committed:
format-rust fix

odzioo123 committed:
small simplification
